### PR TITLE
Clarify that bundle extras should not be used for PyPi installs

### LINKED ISSUE #23692: Conflicts with airflow constraints for airflow 2.3.0 python 3.9
### Apache Airflow version

2.3.0 (latest released)

### What happened

When installing airflow 2.3.0 using pip command with "all" it fails on dependency google-ads

`pip install "apache-airflow[all]==2.3.0" -c "https://raw.githubusercontent.com/apache/airflow/constraints-2.3.0/constraints-3.9.txt"`

> The conflict is caused by:
    apache-airflow[all] 2.3.0 depends on google-ads>=15.1.1; extra == "all"
    The user requested (constraint) google-ads==14.0.0

I changed the version of google-ads to 15.1.1, but then it failed on dependency databricks-sql-connector
> The conflict is caused by:
    apache-airflow[all] 2.3.0 depends on databricks-sql-connector<3.0.0 and >=2.0.0; extra == "all"
    The user requested (constraint) databricks-sql-connector==1.0.2

and then on different dependencies...

### What you think should happen instead

_No response_

### How to reproduce

(venv) [root@localhost]# `python -V`
Python 3.9.7

(venv) [root@localhost]# `pip install "apache-airflow[all]==2.3.0" -c "https://raw.githubusercontent.com/apache/airflow/constraints-2.3.0/constraints-3.9.txt"`

### Operating System

CentOS 7

### Versions of Apache Airflow Providers

_No response_

### Deployment

Virtualenv installation

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [x] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by potiuk**
The bundle extras we have are only used for development and they
should not be used to install airflow from PyPI. This update
to documentation clarifies it.

Closes: #23692

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragement file, named `{pr_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Review by kaxil — APPROVED** [review]

File: docs/apache-airflow/extra-packages-ref.rst [hunk 1]
```diff
@@ -294,7 +294,12 @@ Those are extras that provide support for integration with external systems via
 Bundle extras
 -------------
 
-Those are extras that install one ore more extras as a bundle.
+Those are extras that install one ore more extras as a bundle. Note that those extras should only be used for "development" version
+of Airflow - i.e. when Airflow is installed from sources. Because of the way how bundle extras are constructed they might not
+work when airflow is installed from 'PyPI`.
+
+If you want to install Airflow from PyPI with "all" extras (which should basically be never needed - you almost never need all extras from Airflow),
+you need to list explicitly all the non-bundle extras that you want to install.
```
Comment (eladkal) [inline docs/apache-airflow/extra-packages-ref.rst:302]: Not really relevant here but since we are talking about dev extras...
One of the things that troubles me is that the packages that we bring with `devel` are also part of the constraints that we release to the users - I believe that we should not do that.

For example in https://raw.githubusercontent.com/apache/airflow/constraints-2.3.0/constraints-3.7.txt you can find `freezegun`.

**Review by eladkal — APPROVED** [review]

File: docs/apache-airflow/extra-packages-ref.rst (hunk 1 above)
Comment (kaxil) [reply docs/apache-airflow/extra-packages-ref.rst:302]: Good point, yes

Comment (potiuk) [reply docs/apache-airflow/extra-packages-ref.rst:302]: This is not a problem and removing those devel-only requirements is extremely hard and borderline impossible in automated way. 

**Why this is not a problem?**

Constraints are just constraints not requirements. it's merely a meta-data tellng "if you are installing freezegun, you should be installing this and that version". They have 0 influence on whether freezegun will be installed at all.

**Why this is difficutt or even impossible?**

The problem is that It might be very difficult to separate those out automatically (in a future-safe way) actually because we have no idea if dependency is not used in some of our dependencies beyond our devel deps. 

With Freezegun, this is **probably** fine as you know what freezegun is typically used for, but there were already a few cases where something that "seem" to be only used for devel, was actually used (wrongly or not but there was an implicit dependency) for some packages we installed.

We could posisbly try to figure it out by traversing the dependency tree, but this is doomed because of the way extras are treated (and not even Airflow extras but other packages extras).  The way `pip` treats extras is a bit unexpected (but there is no other good way) - they are optional and only valid during installation and the relation betwen the packages that installed them is gone right after the installation. The fact that there is a dependency between the package and whatever is declared as extra basically disappears at the moment `pip install` completes.

Unfiortunately when a package declares dependency  it MIGHT add extra and that leads to untraceable transitive dependencies. And that dependency - MIGHT  declare other dependencies WITH EXTRAS in THEIR extras, and it can go on and on.

For example: 

if we ourselves (and this is not real case but might happen) add `apache-beam[pandas]` as dependency and `apache -beam[pandas]` adds `pandas[freeze]` as dependency and `pandas[freeze]` adds `freezegun<1.4` as dependency, then we have an implicit, transitive dependency on `freezgun<1.4` from `airflow[apache.beam]`. 

And what's worse beyond the `pip` resolution process, we  have no idea that:

a) freezegun is actually our non-devel transitive dependency (throuhgh beam and pandas)
b) we require freezegun < 1.4

This information is not available anywhere out-of-the-box. In other to know that - we would have to repeat the `pip resolver` process and basically make a snapshot of the in-memory information that teh resolver keeps to find out which packages should be installed. This is the only way to find out that `freezegun` is our non-devel depenedency as well.

If we remove freezegun from our constraints, it might be, that someone installs it in the version 1.5 precisely because we do not have it in constraints (and that will be wrong because `apache-airflow[apache.beam]` requires `freezegun|<1.4`. 

Long story short - freezegun and the above example are pretty artifficial, but there is basically no way (other than repeating what the `pip` resolver does to generate the list of dependencies that are devel-only.

Comment (potiuk) [reply docs/apache-airflow/extra-packages-ref.rst:302]: And yes. Dependencies are HARD :exploding_head:
